Source/Core/Common/RangeSizeSet.h

```diff
@@ -0,0 +1,458 @@
+        return;
+      }
+
+      if (to >= get_to(bound))
+      {
+        if (from == get_from(bound))
+        {
+          // given range fully overlaps last range, erase it
+          erase_range(bound);
+          return;
+        }
+        else
+        {
+          // given range overlaps end of last range, reduce it
+          reduce_to(bound, from);
+          return;
+        }
+      }
+
+      if (from == get_from(bound))
+      {
+        // given range overlaps begin of last range, reduce it
+        reduce_from(bound, to);
+        return;
+      }
+      else
+      {
+        // given range overlaps middle of last range, bisect it
+        bisect_range(bound, from, to);
+        return;
+      }
+    }
+
+    if (bound == m_map.begin())
+    {
+      // if we found the first range that means 'from' is before any stored range
+      // this means we can just erase from start until 'to' and be done with it
+      erase_from_iterator_to_value(bound, to);
+      return;
+    }
+
+    // check previous range
+    auto abound = bound--;
+
+    if (from == get_from(bound))
+    {
+      // similarly, if the previous range starts with the given one, just erase until 'to'
+      erase_from_iterator_to_value(bound, to);
+      return;
+    }
+
+    // if we come here, the given range may or may not overlap part of the current 'bound'
+    // (but never the full range), which means we may need to update the end position of it,
+    // or possibly even split it into two
+    if (from < get_to(bound))
+    {
+      if (to < get_to(bound))
+      {
+        // need to split in two
+        bisect_range(bound, from, to);
+      }
+      else
+      {
+        // just update end
+        reduce_to(bound, from);
+      }
+    }
+
+    // and then just erase until to
+    erase_from_iterator_to_value(abound, to);
+    return;
+  }
+
+  void clear()
+  {
+    m_map.clear();
+    m_sizes.clear();
+  }
+
+  size_t size() const { return m_map.size(); }
+
+  const_iterator begin() const { return const_iterator(m_map.begin()); }
+
+  const_iterator end() const { return const_iterator(m_map.end()); }
+
+  by_size_const_iterator by_size_begin() const { return by_size_const_iterator(m_sizes.begin()); }
+
+  by_size_const_iterator by_size_end() const { return by_size_const_iterator(m_sizes.end()); }
+
+  bool operator==(const RangeSizeSet<T>& other) const { return this->m_map == other.m_map; }
+
+  bool operator!=(const RangeSizeSet<T>& other) const { return !(*this == other); }
+
+private:
+  UT calc_size(T from, T to)
+  {
+    // TODO: This should behave correctly when T is signed.
+    return UT(to - from);
+  }
+
+  // Value type stored in the regular range map.
+  struct Value
+  {
+    // End point of the range.
+    T m_to;
+
+    // Pointer to the same range in the by-size multimap.
+    typename std::multimap<UT, typename std::map<T, Value>::iterator, std::greater<UT>>::iterator
```

Comment:
May want to separate this out into local using aliases, so this is less jarring to read.